Add FIPS support for Android Arm 64-bit [1.0.2]

mspncp opened the pull request:
Fixes #2490
Fixes #8711

In commit 6db8e3bdc9e, support for Android Arm 64-bit was added to the OpenSSL
FIPS Object Module. For some reason, the corresponding target 'android64-aarch64'
was missing OpenSSL 1.0.2, whence it could not be built with FIPS support on
Android Arm 64-bit. This commit adds the missing target.

_(Edited:  new commit message)_

_Note: I am aware that OpenSSL 1.0.2 is in security fix only mode. But maybe missing support in the libcrypto wrapper for a validated FIPS 2.0 object module target can be considered a security fix?_


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte commented:
I agree that this fix is worth an exception from the "security only fixes" policy.

levitte commented on Configure:
```diff
@@ -474,7 +474,7 @@ my %table=(
 "android-x86","gcc:-mandroid -I\$(ANDROID_DEV)/include -B\$(ANDROID_DEV)/lib -O3 -fomit-frame-pointer -Wall::-D_REENTRANT::-ldl:BN_LLONG ${x86_gcc_des} ${x86_gcc_opts}:".eval{my $asm=${x86_elf_asm};$asm=~s/:elf/:android/;$asm}.":dlfcn:linux-shared:-fPIC::.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
 "android-armv7","gcc:-march=armv7-a -mandroid -I\$(ANDROID_DEV)/include -B\$(ANDROID_DEV)/lib -O3 -fomit-frame-pointer -Wall::-D_REENTRANT::-ldl:BN_LLONG RC4_CHAR RC4_CHUNK DES_INT DES_UNROLL BF_PTR:${armv4_asm}:dlfcn:linux-shared:-fPIC::.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
-"android-mips","gcc:-mandroid -I\$(ANDROID_DEV)/include -B\$(ANDROID_DEV)/lib -O3 -Wall::-D_REENTRANT::-ldl:BN_LLONG RC4_CHAR RC4_CHUNK DES_INT DES_UNROLL BF_PTR:${mips32_asm}:o32:dlfcn:linux-shared:-fPIC::.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
+"android64-aarch64","gcc:-mandroid -fPIC -I\$(ANDROID_DEV)/include -B\$(ANDROID_DEV)/lib -O3 -Wall::-D_REENTRANT::-pie%-ldl:SIXTY_FOUR_BIT_LONG RC4_CHAR RC4_CHUNK DES_INT DES_UNROLL BF_PTR:${aarch64_asm}:linux64:dlfcn:linux-shared:::.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
```
Is there a reason why you're removing the android-mips target?

mspncp replied:
Oh, sorry. That was removed accidentally when I compared 1.0.2 with 6db8e3b. I'll revert that.

mspncp replied:
Done, see ca760f0216489f0aa4d17ef25aedd4d5a3a9c2ae.

draak567 replied:
64-bit support on Android also requires x86-64 in addition to AArch64. Not sure if you wanted to add that as well.

levitte replied:
Since this is an import from the old FIPS module, I think it is OK to add android-x86_64 *if* such a thing exists in the FIPS module. Otherwise not.

mspncp replied:
> 64-bit support on Android also requires x86-64 in addition to AArch64. Not sure if you wanted to add that as well.

This pull request intends only to fix compilation for FIPS certified Android 64-bit targets. There is no `android-x86_64` target (or anything similar) in the OpenSSL-fips-2_0-stable branch, only `android-x86` (which is identical to the one in 1.0.2). Also, the FIPS certificates seem to contain only validations for Arm:

https://csrc.nist.gov/projects/cryptographic-module-validation-program/Certificate/1747

```
Android 2.2 running on OMAP 3530 (ARMv7) with NEON (gcc Compiler Version 4.1.0)
Android 2.2 running on Qualcomm QSD8250 (ARMv7) with NEON (gcc Compiler Version 4.4.0)
Android 2.2 running on Qualcomm QSD8250 (ARMv7) without NEON (gcc Compiler Version 4.4.0)
Android 3.0 running on NVIDIA Tegra 250 T20 (ARMv7) (gcc Compiler Version 4.4.0)
Android 4.0 running on NVIDIA Tegra 250 T20 (ARMv7) (gcc Compiler Version 4.4.3)
Android 4.0 running on Qualcomm Snapdragon APQ8060 (ARMv7) with NEON (gcc compiler Version 4.4.3)
Android 4.0 running on TI OMAP 3 (ARMv7) with NEON (gcc Compiler Version 4.4.3)
Android 4.1 running on TI DM3730 (ARMv7) with NEON (gcc Complier Version 4.6)
Android 4.1 running on TI DM3730 (ARMv7) without NEON (gcc Compiler Version 4.6)
Android 4.2 running on Nvidia Tegra 3 (ARMv7) with NEON (gcc Compiler Version 4.6)
Android 4.2 running on Nvidia Tegra 3 (ARMv7) without NEON (gcc Compiler Version 4.6)
Android 5.0 32-bit running on Qualcomm APQ8084 (ARMv7) with NEON (gcc Compiler Version 4.9)
Android 5.0 32-bit running on Qualcomm APQ8084 (ARMv7) without NEON (gcc Compiler Version 4.9)
Android 5.0 64-bit running on SAMSUNG Exynos7420 (ARMv8) with NEON and Crypto Extensions (gcc Compiler Version 4.9) (singleuser mode)
Android 5.0 64-bit running on SAMSUNG Exynos7420 (ARMv8) without NEON and Crypto Extensions (gcc Compiler Version 4.9)
```

https://csrc.nist.gov/projects/cryptographic-module-validation-program/Certificate/2398

```
Android 5.0 32-bit running on Qualcomm APQ8084 (ARMv7) with NEON (gcc Compiler Version 4.9)
Android 5.0 32-bit running on Qualcomm APQ8084 (ARMv7) without NEON (gcc Compiler Version 4.9)
Android 5.0 64-bit running on SAMSUNG Exynos7420 (ARMv8) with NEON and Crypto Extensions (gcc Compiler Version 4.9)
Android 5.0 64-bit running on SAMSUNG Exynos7420 (ARMv8) without NEON and Crypto Extensions (gcc Compiler Version 4.9)
```

mspncp replied:
Note: OpenSSL versions 1.1.0 and higher have full support for Android 64-bit, but they are not compatible with the FIPS 2.0 Object Module.

levitte replied:
That settles it

mspncp commented:
FWIW: incidentally, a colleague of mine is currently struggling with OpenSSL FIPS on Android 64bit, too.

mspncp commented:
Richard, does this need any documentation appart from a CHANGES entry?

levitte commented:
I don't think so.

mspncp commented:
Also, does the CI need to be adjusted to test the build?

levitte commented:
Ah, you should generate a new `TABLE`...  that one's probably lagging behind, so that may be for another PR

levitte commented:
Regarding the CI, we don't have any Android build, so unless you want to work on adding that...  (that's also for another PR, methinks)

mspncp commented:
> Ah, you should generate a new `TABLE`... that one's probably lagging behind, so that may be for another PR

Would you mind doing that for me? (I've never done it before, so I have no experience.)

mspncp commented:
Will if you don't mind that it goes in untested on Android...  Anyway, I will wait for feedback from #8711. (Also, my colleague was able to build successfully with this patch. But we did not run any tests.)

mspncp commented:
> > Ah, you should generate a new `TABLE`... that one's probably lagging behind, so that may be for another PR
> 
> Would you mind doing that for me? (I've never done it before, so I have no experience.)

Ok, updating the TABLE is no problem. Hang on a second...

mspncp commented on TABLE:
```diff
@@ -1,4 +1,3 @@
-Output of `Configure TABLE':
```
Why was this line remove? I ran

```
./Configure TABLE > TABLE
```
Did I miss anything?

mspncp replied:
Restored.

mspncp commented:
@levitte I added a CHANGES entry which is more concise than the previous commit message and adjusted the commit message accordingly.

New commit message is in 2f84f1ddcc42612013d86efaf342bfebec089cc7.

mspncp commented:
This is hopefully my last adjustment. Please reconfirm.

paulidale commented:
FWIW: I'd support an exemption to allow this into 1.0.2.

mspncp commented:
Merged to f937540ec40a5e838460b8f19d2eb722529126b8, thanks!
